src/tir/transforms/common_subexpr_elim_tools.cc

```diff
@@ -0,0 +1,836 @@
+                                                        const TableOfComputations& table2) {
+  TableOfComputations result;
+  for (const auto& current : table1) {
+    auto it = table2.find(current.first);
+    if (it != table2.end()) {
+      result[current.first] = current.second + it->second;
+    }
+  }
+  return result;
+}
+
+/*!
+ * \brief Does the intersection of three tables of computations.
+ * \param table1 One of the three tables, which won't change.
+ * \param table2 One of the three tables, which won't change.
+ * \param table3 One of the three tables, which won't change.
+ * \note We don't need (at least yet) to have a function working for N tables, even if this
+ *       function for 3 tables seems at first glance redundant with the one for 2 tables defined
+ *       just above. The reason is that in order to do the intersection for N tables, we need to
+ *       know how to do it for two. That's because we would compute for N tables using the
+ *       associativity of the intersection : T1 Inter T2 Inter T3 ... Inter Tn
+ *       = ((T1 Inter T2) Inter T3) ... Inter Tn
+ *       Therefore, we need one for 2 tables anyway. And as the one for N (N>=3) would only be used
+ *       (at least for now) for N=3, there is at the moment no need for such a generic intersection
+ *       over N tables.
+ */
+TableOfComputations IntersectionOf3TablesOfComputations(const TableOfComputations& table1,
+                  const TableOfComputations& table2, const TableOfComputations& table3) {
+  TableOfComputations result = IntersectionOf2TablesOfComputations(table1, table2);
+  result = IntersectionOf2TablesOfComputations(result, table3);
+  return result;
+}
+
+/*!
+ * \brief Recompute the number of times that each computation in table_main
+          is being seen in table_bloc1, table_bloc2 and table_bloc3. It sets
+          each element to the sum of the times it is seen in each individual bloc.
+ * \param table_main The main table, for which we recompute the counters.
+ * \param table1 One of the three tables, which won't change.
+ * \param table2 One of the three tables, which won't change.
+ * \param table3 One of the three tables, which won't change.
+ * \note This function is needed because both the intersection (A Inter B) and the union
+ *        (A U B U C) adds the individual counters found in A, B and C. So when we treat for
+ *        instance an If (which contains a Cond, a Then branch and an Else branch),
+ *        it will compute (Then Inter Else) U (Cond Inter Then) U (Cond Inter Else).
+ *        In order to get back to the appropriate number (for instance, 3 if seen one time in each
+ *        bloc), it is therefore necessary to recompute the counters afterwards, which is what this
+ *        function does.
+ */
+void RecomputeNbTimesSeenInThreeBlocs(TableOfComputations& table_main,
+    const TableOfComputations& table_bloc1, const TableOfComputations& table_bloc2,
+    const TableOfComputations& table_bloc3) {
```

Comment:
I prefer taking a vector of `ComputationTable` here and remove `ThreeBlocs` suffix.

Reply:
Good idea! Thanks for the remark!